# 2021-06-20
I have found the cause of the bug I had yesterday, but don't know the solution.
If the randomly generated rules do not include a Antecedent or Consequent then the code breaks because it is expecting all of them to be covered by the rules, and does not know anything about them if not. There are two points of failure:
1. `skfuzzy.control.controlsystem._InputAcceptor.__setitem__` fails if an antecedent in the input dict is not in the rules.
2. `evofuzzy.fuzzyclassifier._make_predictions` argmax fails if a consequent is not covered by the rules.  
 
The second I can do more about since it is in my code - e.g. use classifier.output.get(name, 0) to default to activation of 0.

The first one is harder - how do I handle it?  Some options:
1. add some sort of dummy rule for each antecendent that does some sort of noop.  That may be hard to do in fuzzy logic.
2. hack the code to not throw an exception
3. before running the rule set, check for missing antecedents and remove those columns from the dataframe.  That sounds better and would be more performant.


Fixed the issues - dropped the columns from the dataframe.
I have closed the fitness evaluator story and started on the implement GP engine story.    Completed 21 points out of 62 in 3 weeks.  Not bad.  If I carried on at that rate would finish by 1st August, but am not holding my breath.  